no-dupe-else-if

magurotuna opened the pull request:
~This is WORK IN PROGRESS~

This PR adds [no-dupe-else-if - Rules - ESLint - Pluggable JavaScript linter](https://eslint.org/docs/rules/no-dupe-else-if) rule.

The original implementation is [here](https://github.com/eslint/eslint/blob/master/lib/rules/no-dupe-else-if.js), and it's so complicated.
I did my best to port the original implementation, but the code got dirty, although all the tests are passed.

Also I'm concerned about bad time complexity and cloning `Vec` many times, which may cause bad performance.

bartlomieju commented:
@disizali please review when you get the time

bartlomieju commented on src/rules/no_dupe_else_if.rs:
```diff
@@ -0,0 +1,422 @@
+    assert_lint_err::<NoDupeElseIf>("if (a) {} else if (a && a) {}", 19);
+    assert_lint_err::<NoDupeElseIf>("if (a && a) {} else if (a && a) {}", 24);
+    assert_lint_err::<NoDupeElseIf>("if (a && a) {} else if (a) {}", 24);
+  }
```
One more request from me: please add a test case with nested if:
```
if (something) {
  if (a == 1) {
  } else if (a == 2) {}
}
```

magurotuna commented on src/rules/no_dupe_else_if.rs:
```diff
@@ -0,0 +1,422 @@
+    | (Unary(_), Unary(_))
+    | (Update(_), Update(_))
+    | (Bin(_), Bin(_))
+    | (Assign(_), Member(_))
```
Oops, I made a mistake! I'll fix it soon.

magurotuna commented:
@disizali @bartlomieju 
Thanks for the review. I've added more tests and fixed the implementation in `equal_in_if_else`.
Could you review again please.
